### ISSUE TITLE
WFC3 Saturation map code release

### ISSUE THREAD

**Issue by stscijgbot-hstdp** [spacetelescope/drizzlepac#1377]
_Issue [HLA-776](https://jira.stsci.edu/browse/HLA-776) was created on JIRA by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena):_

Address any issues found by the WFC3 team during testing, and then prepare the code for release.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=590842#comment-590842):_

Additional tickets may need to be written if specific code changes need to be done.  At the least, new truth files need to be generated for the WFC3 PyTests.

**Comment by stsci-hack** [comment]
This issue relates to code outside this package.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=590824#comment-590824):_

This work is pending thorough evaluation of the updated CALWF3 code in conjunction with the new saturation map calibration reference files.  WFC3 team point of contact, Jenn Mack, indicated (on 02 June 2022) this evaluation and testing may not be done for months.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=590842#comment-590842):_

Additional tickets may need to be written if specific code changes need to be done.  At the least, new truth files need to be generated for the WFC3 PyTests.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=637986#comment-637986):_

Sent to Isabel:

Good Afternoon,

I am glad to know the software, calwf3, worked for you!  My comments and questions are below.

Thank You in advance and Best,

Michele
 # You indicate there seems to be a difference in terms of use of “>” (single CCDTable value) versus “>=” (map image).  Was this an empirical determination based on your examination of the output data?  I ask because the actual code is below where the comparison is using “>” just like what was done for the single value.  You have already indicated your results are “optimal” so this is really a follow-through question for me. ☺

            for (i = xbeg[0];  i < xend[0];  i++) {

                /* Flag full-well saturated pixels with 256 dq bit */

                if (Pix(x->sci.data, i, j) > (Pix(satimage.sci.data, i, j) / wf3->mean_gain)) {

                    sum_dq = DQPix(x->dq.data, i, j) | SATPIXEL;

                    DQSetPix(x->dq.data, i, j, sum_dq);

                }

            }}
 # Sorry to be dense, but I do not understand what you mean by {_}mixed rows and columns{_}.  Are you referring to the overscan regions?  These regions should be ignored by the software.
 # From my perspective, all calibration/reference data is controlled by the Instrument Teams.  I just helped out with the Saturation images because they are a new file.  While I developed the code, I could say what keywords would be needed in the header to support the new processing.  At this time, these new files contain everything that is needed to support CALWF3.  If WFC3 Team wants to add more keywords to the headers to provide more information, WFC3 Team should feel empowered to do so!
 # While I have processed some data, for my final test I would like to have one or two RAW and the corresponding FLC files from your testing.  This is a way for me to ensure that we are using the same software and getting the same results.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=639001#comment-639001):_

Tested with my own test datasets, and more importantly, with datasets provided by WFC3 (1x1: iem354khq, 2x2: ibl999tfqa, 3x3: iacs02tyq, and subarray: if2r02agq. My results agree with those of WFC3.  The code is working properly.  I did update the "date" of the work done in Dates, History, and Updates to the current date so that it more closely corresponds to the actual build release date.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=637986#comment-637986):_

Sent to Isabel:

Good Afternoon,

I am glad to know the software, calwf3, worked for you!  My comments and questions are below.

Thank You in advance and Best,

Michele
 # You indicate there seems to be a difference in terms of use of “>” (single CCDTable value) versus “>=” (map image).  Was this an empirical determination based on your examination of the output data?  I ask because the actual code is below where the comparison is using “>” just like what was done for the single value.  You have already indicated your results are “optimal” so this is really a follow-through question for me. ☺

            for (i = xbeg[0];  i < xend[0];  i++) {

                /* Flag full-well saturated pixels with 256 dq bit */

                if (Pix(x->sci.data, i, j) > (Pix(satimage.sci.data, i, j) / wf3->mean_gain)) {

                    sum_dq = DQPix(x->dq.data, i, j) | SATPIXEL;

                    DQSetPix(x->dq.data, i, j, sum_dq);

                }

            }}
 # Sorry to be dense, but I do not understand what you mean by {_}mixed rows and columns{_}.  Are you referring to the overscan regions?  These regions should be ignored by the software.
 # From my perspective, all calibration/reference data is controlled by the Instrument Teams.  I just helped out with the Saturation images because they are a new file.  While I developed the code, I could say what keywords would be needed in the header to support the new processing.  At this time, these new files contain everything that is needed to support CALWF3.  If WFC3 Team wants to add more keywords to the headers to provide more information, WFC3 Team should feel empowered to do so!
 # While I have processed some data, for my final test I would like to have one or two RAW and the corresponding FLC files from your testing.  This is a way for me to ensure that we are using the same software and getting the same results.

**Comment by stscijgbot-hstdp** [comment]
_Comment by [Michele De La Pena](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mdelapena) on [JIRA](https://jira.stsci.edu/browse/HLA-776?focusedCommentId=639001#comment-639001):_

Tested with my own test datasets, and more importantly, with datasets provided by WFC3 (1x1: iem354khq, 2x2: ibl999tfqa, 3x3: iacs02tyq, and subarray: if2r02agq. My results agree with those of WFC3.  The code is working properly.  I did update the "date" of the work done in Dates, History, and Updates to the current date so that it more closely corresponds to the actual build release date.
